File: intel_briefing/sensors/hacker_news.py
# ABOUTME: Hacker News sensor using the official Firebase API.
# ABOUTME: Fetches top stories with heat scores and timestamps.
"""
Hacker News Sensor - Fetches top stories from Hacker News.
Uses the official Firebase API (no auth required).
"""
import sys
import json
from dataclasses import dataclass
from typing import List, Optional

import httpx

@dataclass
class HNStory:
    """A Hacker News story."""
    id: int
    title: str
    url: Optional[str]
    score: int
    by: str
    descendants: int  # comment count
    
    @property
    def hn_url(self) -> str:
        return f"https://news.ycombinator.com/item?id={self.id}"

def fetch_top_stories(limit: int = 10) -> List[HNStory]:
    """Fetch top stories from Hacker News."""
    print(f"  → Fetching top {limit} stories from Hacker News...")
    
    # Get top story IDs
    resp = httpx.get("https://hacker-news.firebaseio.com/v0/topstories.json", timeout=15)
    story_ids = resp.json()[:limit]
    
    stories = []
    for sid in story_ids:
        item_resp = httpx.get(f"https://hacker-news.firebaseio.com/v0/item/{sid}.json", timeout=10)
        item = item_resp.json()
        if item and item.get("type") == "story":
            stories.append(HNStory(
                id=item["id"],
                title=item.get("title", ""),
                url=item.get("url"),
                score=item.get("score", 0),
                by=item.get("by", "unknown"),
                descendants=item.get("descendants", 0)
            ))
    
    return stories

def print_stories(stories: List[HNStory]):
    """Print stories in a readable format."""
    print(f"\n{'='*60}")
    print(f"  📰 Hacker News Top Stories")
    print(f"{'='*60}\n")
    
    for i, s in enumerate(stories, 1):
        print(f"{i}. {s.title}")
        print(f"   ⬆️ {s.score} points | 💬 {s.descendants} comments | by {s.by}")
        print(f"   🔗 {s.url or s.hn_url}")
        print()

if __name__ == "__main__":
    limit = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    stories = fetch_top_stories(limit)
    if stories:
        print_stories(stories)
    else:
        print("No stories found.")

File: frontend/src/lib/sensors/hacker_news.ts
// ABOUTME: Hacker News sensor using the official Firebase REST API.
// ABOUTME: Fetches top stories with parallel fetching and top-level comments.
import type { ConfigSettings, IntelItem } from '../models'

const HN_BASE = 'https://hacker-news.firebaseio.com/v0'
const MAX_COMMENTS = 5

function stripHtml(html: string): string {
  return html.replace(/<[^>]+>/g, '').trim()
}

interface HnComment {
  by?: string
  text?: string
}

async function fetchComments(kids: number[]): Promise<string> {
  const commentIds = kids.slice(0, MAX_COMMENTS)
  const settled = await Promise.allSettled(
    commentIds.map(async (id) => {
      const resp = await fetch(`${HN_BASE}/item/${id}.json`, {
        signal: AbortSignal.timeout(10000),
      })
      if (!resp.ok) return null
      return (await resp.json()) as HnComment
    }),
  )

  const lines: string[] = []
  for (const result of settled) {
    if (result.status !== 'fulfilled' || !result.value) continue
    const { by, text } = result.value
    if (!by || !text) continue
    const clean = stripHtml(text).slice(0, 200)
    lines.push(`@${by}: ${clean}`)
  }

  return lines.length > 0 ? `Top comments:\n${lines.join('\n')}` : ''
}

async function fetchStory(storyId: number): Promise<IntelItem | null> {
  try {
    const itemResp = await fetch(`${HN_BASE}/item/${storyId}.json`, {
      signal: AbortSignal.timeout(10000),
    })
    if (!itemResp.ok) return null
    const item = (await itemResp.json()) as Record<string, unknown>
    if (!item || item.type !== 'story') return null

    const score = Number(item.score ?? 0)
    const descendants = Number(item.descendants ?? 0)
    const url = (item.url as string) || `https://news.ycombinator.com/item?id=${storyId}`
    const time = Number(item.time ?? 0)
    const kids = (item.kids as number[]) ?? []

    const content = kids.length > 0 ? await fetchComments(kids) : ''

    return {
      id: `hn-${storyId}`,
      source: 'hacker_news',
      title: String(item.title ?? ''),
      url,
      heat: `${score} pts, ${descendants} comments`,
      published_at: time > 0 ? new Date(time * 1000).toISOString() : null,
      content: content || null,
    }
  } catch {
    return null
  }
}

export async function fetchHackerNews(_config: ConfigSettings, limit: number): Promise<IntelItem[]> {
  try {
    const resp = await fetch(`${HN_BASE}/topstories.json`, {
      signal: AbortSignal.timeout(15000),
    })
    if (!resp.ok) throw new Error(`HTTP ${resp.status} from Hacker News`)
    const storyIds = (await resp.json()) as number[]

    const candidateIds = storyIds.slice(0, Math.min(limit * 2, 30))

    const settled = await Promise.allSettled(
      candidateIds.map((id) => fetchStory(id)),
    )

    const items: IntelItem[] = []
    for (const result of settled) {
      if (items.length >= limit) break
      if (result.status === 'fulfilled' && result.value) {
        items.push(result.value)
      }
    }
    return items
  } catch (err) {
    throw err instanceof Error ? err : new Error(String(err))
  }
}
